Art/liquidate v2 rebase edition

whenceforth opened the pull request:
a basic liquidation test is working in MoneyMarket.js.

hayesgm commented on contracts/Borrower.sol:
```diff
@@ -341,20 +341,31 @@ contract Borrower is Graceful, Owned, Ledger {
             return false;
         }
         // TODO valid if borrower has loans for multiple assets?
```
This should be correct, right, since `valueEq` takes that into account?

whenceforth replied:
I think valueEq is not useful and we should instead always just compare the absolute value of supply to absolute value of borrows. Consider the following:

Suppose you have 1000 equiv supply. You borrow 2000 equiv.
You can becuase your net value is 1000 and 1000 * ratio 2 = 2000.

Suppose instead you only borrow 1000 equiv.
Now your valueEq becomes 1000supply - 1000loan = 0.
Can you take out a 2nd 1000 equiv borrow to bring your total borrows up to 2000? No.
0 * 2 = 0 < 1000.

In both cases, the desired total borrow amount is 2000 equiv, but you can't do it in 2 chunks.

hayesgm commented on contracts/Borrower.sol:
```diff
@@ -341,20 +341,31 @@ contract Borrower is Graceful, Owned, Ledger {
 
-        int256 requiredValue = int256(ve.borrowValue / borrowStorage.minimumCollateralRatio());
-        return requiredValue - netValueEquivalent;
+        // Example: 1000 borrows value / 2 ratio = 500 requiredSupplyValue
```
... isn't this backwards? If we have 1,000 borrows value, we require 2000 supply value before we can begin to liquidate.

whenceforth replied:
up in validCollateralRatio, we take the borrow balance and multiply by the ratio to get the max loan.  the ratio is 2, so if you have 500 supply, you can borrow 1000. That means we are only 50% collateralized. 

This method reflects the same logic.

hayesgm commented on contracts/Borrower.sol:
```diff
@@ -50,6 +57,23 @@ contract Borrower is Graceful, Owned, Ledger {
+        if (!checkOwner()) {
+            return false;
+        }
+        if(basisPoints > 3000) {
```
Can we move this to a constant?

whenceforth replied:
will do

hayesgm commented on contracts/Borrower.sol:
```diff
@@ -114,61 +139,115 @@ contract Borrower is Graceful, Owned, Ledger {
+        // current liquidation is completed.
+
+        // How much collateral should the liquidator receive?
+        uint256 seizeCollateralAmount =
```
I believe this is covered, but if I try to seize a random asset that has no value (say they deposited `rando_coin`, this would disallow them here?)

whenceforth replied:
yeah, if the collateral goes to zero, the liquidator WILL NOT be able to seize any of it.  We could special case that, but not sure it's worth it.   First thing that comes to mind: If price is zero, set seizable amount to the whole borrow balance.  Might be a good Robert question.

hayesgm replied:
Well, we can always avoid putting it to precisely zero (though, maybe it's out-of-date, not zero). I was referring to what you were seizing. My fear is that we don't have a price set for EOS tokens, the borrower has them, and so the liquidator takes all the EOS for free. I believe it's not possible, just wanted to check.

hayesgm commented on contracts/Borrower.sol:
```diff
@@ -239,19 +322,70 @@ contract Borrower is Graceful, Owned, Ledger {
+        ValueEquivalents memory ve = getValueEquivalents(borrower);
+
+        // Example: 1000 borrows value / 2 ratio = 500 requiredSupplyValue
+        uint256 requiredSupplyValue = ve.borrowValue / borrowStorage.minimumCollateralRatio();
```
Just to say, this logic was backwards and is now fixed? If you have 1,000 eth, you can borrow 500 eth-worth of something, not the other way around. Otherwise, you'd borrow 2,000 eth and then just walk away.

whenceforth replied:
I will make a separate PR merging onto this one in which that backwardness is fixed

hayesgm commented on test/MoneyMarket.js:
```diff
@@ -227,6 +230,62 @@ contract('MoneyMarket', function(accounts) {
+
+      // Make pig token cheap so loan is easy to get
+      await priceOracle.setAssetValue(faucetToken.address, toAssetValue(1) , {from: web3.eth.accounts[system]});
+      await moneyMarket.customerBorrow(faucetToken.address, 200, {from: web3.eth.accounts[borrower]});
```
If we haven't set the asset value for eth yet, how are we able to borrow here?

whenceforth replied:
It gets set in the before_each block on line 88.
`await utils.setAssetValue(priceOracle, etherToken, 1, web3);`

hayesgm commented on test/MoneyMarket.js:
```diff
@@ -227,6 +230,62 @@ contract('MoneyMarket', function(accounts) {
+      await priceOracle.setAssetValue(faucetToken.address, toAssetValue(1500) , {from: web3.eth.accounts[system]});
+
+      const result = await moneyMarket.liquidateCollateral.call(web3.eth.accounts[borrower], faucetToken.address, 50, etherToken.address, {from: web3.eth.accounts[liquidator]});
+      assert.equal(result, 76530);
```
Where is this number coming from? `discount * 1500 * 200 / 1 = 30000 * discount`?

whenceforth replied:
I've changed this to clarify:
`
const result = await moneyMarket.liquidateCollateral.call(web3.eth.accounts[borrower], faucetToken.address, 50, etherToken.address, {from: web3.eth.accounts[liquidator]});
      const expected_collateral_seized = 76530; // (1500 price * 50quantity) / .98 discount = 75,000 / 0.98 = 76,530.6122449 = 76530 wei weth
      assert.equal(result, expected_collateral_seized);
`

hayesgm commented on test/MoneyMarket.js:
```diff
@@ -320,5 +379,31 @@ contract('MoneyMarket', function(accounts) {
+        const balance = await moneyMarket.getConvertedAssetValueWithDiscount.call(tokenAddrs.BAT, (10 ** 18), tokenAddrs.OMG, 500);
+        // compare to 400000000000000000 in non-discounted test in priceOracle.js of getConvertedAssetValue
+        // we expect to get more of the target asset here because its price has been discounted
+        assert.equal(balance.valueOf(), 421052631578947368); // (1 * 10^18)*2/(5*.95) or 4.444....e17
```
This is `discount * price * amount / otherPrice`, correct?

whenceforth replied:
I've updated code to clarify: 
```javascript
expected: ( quantity 1 * 10^18)*(2 price BAT)/(5 price OMG *.95 discount multiplier) = 4.21052631578947368... × 10^17
```

hayesgm commented on contracts/Borrower.sol:
```diff
@@ -229,8 +309,12 @@ contract Borrower is Graceful, Owned, Ledger {
+        if(valueEquivalent <= 0) {
+            return 0;
+        }
+        return uint256(valueEquivalent) * borrowStorage.minimumCollateralRatio();
```
This is clearly wrong. We should fix and add a test case that verifies it.

whenceforth commented:
@hayesgm I've extracted a CollateralCalculator so both Borrow and Withdraw can use it.  It's hooked up for borrow, but I'm still getting tests to work on the withdraw side.

whenceforth commented on test/Wallet.js:
```diff
@@ -208,76 +208,74 @@ contract('Wallet', function(accounts) {
-      assert.equal(await utils.ledgerAccountBalance(moneyMarket, wallet.address, etherToken.address), 55000000000000000);
+      assert.equal(await utils.ledgerAccountBalance(moneyMarket, wallet.address, etherToken.address), 55000);
 
-      // TODO: This should fail at 27.5, not 110. Check we're calculating ratios correctly.
```
woot! TODO is DONE.

whenceforth commented:
@hayesgm you should probably have another look.  This is ready to review, modulo migrations.  If we AREN'T resetting prod truffle filesystem, I'll need to add some sort of migration. 

key changes: 

- Extracted CollateralCalculator from Borrower
- Gating withdrawals based on collateral requirements by using the extending the collateral calculator in Supplier.sol
- Collateral ratio applied in proper direction
- Moved collateral ratio config to the calculator.  
- Required collateral calculated using the simple formulas from our discussion of 2/20

This will probably put MoneyMarket over the etherscan limit, but I wanted to get a working system together first before attempting any switch to composition over inheritance.

whenceforth commented:
![ethereum solidity contract source code verification bytecode abi 2018-02-21 13-35-54](https://user-images.githubusercontent.com/687237/36506732-52d6e41c-170c-11e8-9f43-9d16c1b48b3a.png)

hayesgm commented on contracts/CollateralCalculator.sol:
```diff
@@ -0,0 +1,164 @@
+            return false;
+        }
+
+        // enforce non-zero input and de-scaled value >= 1
```
This should probably be strictly greater than 1

whenceforth replied:
yes. will fix.

hayesgm commented on contracts/CollateralCalculator.sol:
```diff
@@ -0,0 +1,164 @@
+        ValueEquivalents memory ve = getValueEquivalents(account);
+
+        uint256 ratioAdjustedBorrow = (ve.borrowValue * scaledMinCollateralToBorrowRatio)/ collateralRatioScale;
+        //failure("DEBUG::getMaxWithdrawAvailable: ratioAdjustedBorrow, ve.supplyValue", ratioAdjustedBorrow, ve.supplyValue);
```
comments

hayesgm commented on contracts/CollateralCalculator.sol:
```diff
@@ -0,0 +1,164 @@
+      * @param asset proposed for withdrawal
+      * @param amount amount of asset proposed for withdrawal
+      */
+    function canWithdrawCollateral(address account, address asset, uint256 amount) public returns (bool) {
```
Yay!! 🎆 🍾

hayesgm commented on contracts/CollateralCalculator.sol:
```diff
@@ -0,0 +1,164 @@
+      * @param borrowAsset denomination of borrow
+      * @return boolean true if the requested amount is acceptable and false otherwise
+      */
+    function canBorrowAssetAmount(address borrower, uint256 borrowAmount, address borrowAsset) internal returns (bool) {
```
Yay!! ❤️ 🎊

hayesgm commented on contracts/CollateralCalculator.sol:
```diff
@@ -0,0 +1,164 @@
+        uint256 supplyValue = 0;
+        uint256 borrowValue = 0;
+
+        for (uint64 i = 0; i < assetCount; i++) {
```
Just in case we ever support more than 6.4 billion asset types.

hayesgm replied:
(making fun of the 64-bit iterater)

whenceforth replied:
for the record, I didn't make that uint64 declaration... and at the rate crypto is moving, 6.4 billion asset types seems achievable ;-) but I've changed it to uint8.

whenceforth commented on contracts/Borrower.sol:
```diff
@@ -114,61 +122,115 @@ contract Borrower is Graceful, Owned, Ledger {
+    function liquidateCollateral(address borrower, address borrowedAsset, uint borrowedAssetAmount, address assetToLiquidate) public returns (uint256) {
 
-        return true;
+        uint256 liquidationAmount = previewLiquidateCollateral(borrower, borrowedAsset, borrowedAssetAmount, assetToLiquidate);
```
we have a `borrowedAsset` variable. 
will rename `assetToLiquidate` as `collateralAsset`

whenceforth commented on contracts/Borrower.sol:
```diff
@@ -114,61 +122,115 @@ contract Borrower is Graceful, Owned, Ledger {
-            failure("Borrower::FailedToSaveBlockInterest", uint256(borrowAsset), uint256(LedgerAccount.Borrow));
-            return false;
+        // Make sure msg.sender has as much of the borrowed asset as they are claiming to deliver
+        // NOTE: liquidator is responsible for truing up their balance with any accrued interest, if that is necessary
```
we can't actually call Supplier.accrueSupplyInterest in Borrower.sol... unless we add a setSupplier to Borrower (which seems a bit much) or do some other sort of refactor.

whenceforth commented on contracts/Borrower.sol:
```diff
@@ -224,53 +286,47 @@ contract Borrower is Graceful, Owned, Ledger {
-        uint256 balance = 0;
+    function getConvertedAssetValueWithDiscount(address srcAsset, uint256 srcAssetAmount, address targetAsset, uint16 targetDiscountRateBPS) public view returns(uint) {
+
+        if(srcAsset == targetAsset) {
```
good catch. i'm returning a totally bogus value here.   will return 0 instead.

whenceforth commented on contracts/Borrower.sol:
```diff
@@ -224,53 +286,47 @@ contract Borrower is Graceful, Owned, Ledger {
-        for (uint64 i = 0; i < assetCount; i++) {
-          address asset = priceOracle.assets(i);
+        // We get scaled value from oracle and add discount scaling on top.
+        uint scaledSrcValue = scaledDiscountPrice(priceOracle.getScaledValue(srcAsset), 0);
```
maybe...but we need some scaling so a small discount works properly.   i'll reduce the scaling from 10**9 to 10**5.

whenceforth commented:
This is ready to go after build.

whenceforth commented:
latest version still works with etherscan

hayesgm commented:
Feel free to merge
